eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs

```diff
@@ -64,87 +67,249 @@ import Cardano.Binary
 import Cardano.Ledger.TxIn (TxIn (..))
 import Cardano.Ledger.Val
   ( DecodeNonNegative,
+    Val (..),
+    adaOnly,
     decodeMint,
     decodeNonNegative,
     encodeMint,
     isZero,
   )
+import Data.Bits
+import qualified Data.ByteString as BS
 import Data.Coders
-import Data.Maybe (fromMaybe)
+import Data.Maybe (fromJust, fromMaybe)
 import Data.MemoBytes (Mem, MemoBytes (..), memoBytes)
 import Data.Sequence.Strict (StrictSeq)
 import qualified Data.Sequence.Strict as StrictSeq
 import Data.Set (Set)
 import qualified Data.Set as Set
-import Data.Typeable (Typeable)
+import Data.Typeable (Proxy (..), Typeable, (:~:) (Refl))
+import Data.Word
 import GHC.Generics (Generic)
 import GHC.Records (HasField (..))
 import GHC.Stack (HasCallStack)
+import GHC.TypeLits
 import NoThunks.Class (InspectHeapNamed (..), NoThunks)
 import Prelude hiding (lookup)
 
 data TxOut era
-  = TxOutCompact
+  = TxOutCompact'
       {-# UNPACK #-} !(CompactAddr (Crypto era))
       !(CompactForm (Core.Value era))
-  | TxOutCompactDH
+  | TxOutCompactDH'
       {-# UNPACK #-} !(CompactAddr (Crypto era))
       !(CompactForm (Core.Value era))
       !(DataHash (Crypto era))
+  | SizeHash (CC.ADDRHASH (Crypto era)) ~ 28 =>
+    TxOut_AddrHash28_AdaOnly
+      !(StakeReference (Crypto era))
+      {-# UNPACK #-} !Word64 -- Payment Addr
+      {-# UNPACK #-} !Word64 -- Payment Addr
+      {-# UNPACK #-} !Word64 -- Payment Addr
+      {-# UNPACK #-} !Word64 -- Payment Addr (32bits) + ... +  0/1 for Testnet/Mainnet + 0/1 Script/Pubkey
+      {-# UNPACK #-} !Word64 -- Ada value
+  | ( SizeHash (CC.ADDRHASH (Crypto era)) ~ 28,
+      SizeHash (CC.HASH (Crypto era)) ~ 32
+    ) =>
+    TxOut_AddrHash28_AdaOnly_DataHash32
+      !(StakeReference (Crypto era))
+      {-# UNPACK #-} !Word64 -- Payment Addr
+      {-# UNPACK #-} !Word64 -- Payment Addr
+      {-# UNPACK #-} !Word64 -- Payment Addr
+      {-# UNPACK #-} !Word64 -- Payment Addr (32bits) + ... +  0/1 for Testnet/Mainnet + 0/1 Script/Pubkey
+      {-# UNPACK #-} !Word64 -- Ada value
+      {-# UNPACK #-} !Word64 -- DataHash
+      {-# UNPACK #-} !Word64 -- DataHash
+      {-# UNPACK #-} !Word64 -- DataHash
+      {-# UNPACK #-} !Word64 -- DataHash
 
 deriving stock instance
   ( Eq (Core.Value era),
     Compactible (Core.Value era)
   ) =>
   Eq (TxOut era)
 
+decodeAddress28 ::
+  forall crypto.
+  ( SizeHash (CC.ADDRHASH crypto) ~ 28,
+    HashAlgorithm (CC.ADDRHASH crypto)
+  ) =>
+  StakeReference crypto ->
+  Word64 ->
+  Word64 ->
+  Word64 ->
+  Word64 ->
+  Addr crypto
+decodeAddress28 stakeRef a b c d =
+  Addr network paymentCred stakeRef
+  where
+    network = if d `testBit` 1 then Mainnet else Testnet
+    paymentCred =
+      if d `testBit` 0
+        then KeyHashObj (KeyHash addrHash)
+        else ScriptHashObj (ScriptHash addrHash)
+    addrHash :: Hash (CC.ADDRHASH crypto) a
+    addrHash =
+      hashFromPackedBytes $
+        PackedBytes32 a b c (fromIntegral (d `shiftR` offset))
```

Comment:
I'll be really surprised if this compiles. `PackedBytes32` is a 32 byte hash constructor, but we are decoding an address with 28 bytes. Also `offset` doesn't seem to be defined